### ISSUE TITLE
Inconsistent inference with batch size=1

### ISSUE THREAD

**Issue by dimiz51** [pytorch/pytorch#147700]
### 🐛 Describe the bug

I have built a custom implementation of DETR (Detection transformer) using a ResNet50 backbone from torchvision. The problem is that inference fails to produce any detections with batch size = 1 while it works fine for any other batch size > 1.

**Forward pass of my model**
```
    def forward(self, x):
        # Pass inputs through the CNN backbone...
        tokens = self.backbone(x)["layer4"]

        # Pass outputs from the backbone through a simple conv...
        tokens = self.conv1x1(tokens)

        # Re-order in patches format
        tokens = rearrange(tokens, "b c h w -> b (h w) c")

        # Pass encoded patches through encoder...
        out_encoder = self.transformer_encoder((tokens + self.pe_encoder))

        # We expand so each image of each batch get's it's own copy of the
        # query embeddings. So from (1, 100, 256) to (4, 100, 256) for example
        # for batch size=4, with 100 queries of embedding dimension 256.
        queries = self.queries.repeat(out_encoder.shape[0], 1, 1)

        # Compute outcomes for all intermediate
        # decoder's layers...
        class_preds = []
        bbox_preds = []

        for layer in self.transformer_decoder.layers:
            queries = layer(queries, out_encoder)
            class_preds.append(self.linear_class(queries))
            bbox_preds.append(self.linear_bbox(queries))

        # Stack and return
        class_preds = torch.stack(class_preds, dim=1)
        bbox_preds = torch.stack(bbox_preds, dim=1)

        return class_preds, bbox_preds
```

**My inference code**
```
def run_inference(
    model,
    device,
    inputs,
    nms_threshold=0.3,
    image_size=480,
    empty_class_id=0,
    out_format="xyxy",
    scale_boxes=True,
):
    """
    Utility function that wraps the inference and post-processing and returns the results for the
    batch of inputs. The inference will be run using the passed model and device while post-processing
    will be done on the CPU.

    Args:
        model (torch.nn.Module): The trained model for inference.
        device (torch.device): The device to run inference on.
        inputs (torch.Tensor): Batch of input images.
        nms_threshold (float, optional): NMS threshold for removing overlapping boxes. Default is 0.3.
        image_size (int, optional): Image size for transformations. Default is 480.
        empty_class_id (int, optional): The class ID representing 'no object'. Default is 0.
        out_format (str, optional): Output format for bounding boxes. Default is "xyxy".
        scale_boxes (bool, optional): Whether to scale the bounding boxes. Default is True.
    Returns:
        List of tuples: Each tuple contains (nms_boxes, nms_probs, nms_classes) for a batch item.
    """
    if model and device:
        model.eval()
        model.to(device)
        inputs = inputs.to(device)
    else:
        raise ValueError("No model or device provided for inference!")

    with torch.no_grad():
        out_cl, out_bbox = model(inputs)

    # Get the outputs from the last decoder layer..
    out_cl = out_cl[:, -1, :]
    out_bbox = out_bbox[:, -1, :]
    out_bbox = out_bbox.sigmoid().cpu()
    out_cl_probs = out_cl.cpu()

    scale_factors = torch.tensor([image_size, image_size, image_size, image_size])
    results = []

    for i in range(inputs.shape[0]):
        o_bbox = out_bbox[i]
        o_cl = out_cl_probs[i].softmax(dim=-1)
        o_bbox = ops.box_convert(o_bbox, in_fmt="cxcywh", out_fmt=out_format)

        # Scale boxes if needed...
        if scale_boxes:
            o_bbox = o_bbox * scale_factors

        # Filter out boxes with no object...
        o_keep = o_cl.argmax(dim=-1) != empty_class_id
        if o_keep.sum() == 0:
            results.append((np.array([]), np.array([]), np.array([])))
            continue
        keep_probs = o_cl[o_keep]
        keep_boxes = o_bbox[o_keep]

        # Apply NMS
        nms_boxes, nms_probs, nms_classes = class_based_nms(
            keep_boxes, keep_probs, nms_threshold
        )
        results.append((nms_boxes, nms_probs, nms_classes))

    return results
```

**Note:** Probabilities after softmax with batch size 1 are always near zero for any other class than the "empty" class (ID=0).

Any ideas?

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.13 (main, Nov  9 2023, 01:24:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-52-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        43 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               8
On-line CPU(s) list:                  0-7
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 5 2500U with Radeon Vega Mobile Gfx
CPU family:                           23
Model:                                17
Thread(s) per core:                   2
Core(s) per socket:                   4
Socket(s):                            1
Stepping:                             0
Frequency boost:                      enabled
CPU max MHz:                          2000.0000
CPU min MHz:                          1600.0000
BogoMIPS:                             3992.31
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev sev_es
Virtualization:                       AMD-V
L1d cache:                            128 KiB (4 instances)
L1i cache:                            256 KiB (4 instances)
L2 cache:                             2 MiB (4 instances)
L3 cache:                             4 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-7
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; untrained return thunk; SMT vulnerable
Vulnerability Spec rstack overflow:   Mitigation; Safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.1
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] onnx==1.16.1
[pip3] onnxruntime==1.18.0
[pip3] optree==0.11.0
[pip3] torch==2.6.0
[pip3] torchinfo==1.8.0
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.21.0
[pip3] triton==3.2.0
[conda] Could not collect

**Comment by hakankocakk** [comment]
You may use Weighted Cross-Entropy Loss and Focal loss. I would be happy if you share the result with me.

**Comment by dimiz51** [comment]
> You may use Weighted Cross-Entropy Loss and Focal loss. I would be happy if you share the result with me.

I currently use Cross entropy during training. Can you elaborate how you think that affects inference related to my batch size issue? Not sure I understand..As mentioned, results are quite decent with any other batch size !=1. I think it might have something to do with batch norms related to the Transformer/Backbone layers I use but not sure at all.

**Model code for reference**
```

class DETR(nn.Module):
    """Detection Transformer (DETR) model with a ResNet50 backbone.

    Paper: https://arxiv.org/abs/2005.12872

    Args:
        d_model (int, optional): Embedding dimension. Defaults to 256.
        n_classes (int, optional): Number of classes. Defaults to 92.
        n_tokens (int, optional): Number of tokens. Defaults to 225.
        n_layers (int, optional): Number of layers. Defaults to 6.
        n_heads (int, optional): Number of heads. Defaults to 8.
        n_queries (int, optional): Number of queries/max objects. Defaults to 100.

    Returns:
        DETR: DETR model
    """

    def __init__(
        self,
        d_model=256,
        n_classes=92,
        n_tokens=225,
        n_layers=6,
        n_heads=8,
        n_queries=100,
    ):
        super().__init__()

        self.backbone = create_feature_extractor(
            torch.hub.load("pytorch/vision:v0.10.0", "resnet50", pretrained=True),
            return_nodes={"layer4": "layer4"},
        )

        self.conv1x1 = nn.Conv2d(2048, d_model, kernel_size=1, stride=1)

        self.pe_encoder = nn.Parameter(
            torch.rand((1, n_tokens, d_model)), requires_grad=True
        )

        encoder_layer = nn.TransformerEncoderLayer(
            d_model=d_model,
            nhead=n_heads,
            dim_feedforward=4 * d_model,
            dropout=0.1,
        )

        self.transformer_encoder = nn.TransformerEncoder(
            encoder_layer, num_layers=n_layers
        )

        self.queries = nn.Parameter(
            torch.rand((1, n_queries, d_model)), requires_grad=True
        )

        decoder_layer = nn.TransformerDecoderLayer(
            d_model=d_model,
            nhead=n_heads,
            dim_feedforward=4 * d_model,
            batch_first=True,
            dropout=0.1,
        )

        self.transformer_decoder = nn.TransformerDecoder(
            decoder_layer, num_layers=n_layers
        )

        # Each of the decoder's outputs will be passed through
        # linear layers for prediction of boxes/classes.
        self.linear_class = nn.Linear(d_model, n_classes)
        self.linear_bbox = nn.Linear(d_model, 4)

        # Add hooks to get intermediate outcomes
        self.decoder_outs = {}
        for i, L in enumerate(self.transformer_decoder.layers):
            name = f"layer_{i}"
            L.register_forward_hook(get_hook(self.decoder_outs, name))
```

**Comment by dimiz51** [comment]
**UPDATE**: Seems like Batch norms might indeed be the issue, i calculated the mean/std over training and validation dataset(batch size 1 here) and those tend to differ a bit...Here is a few things i tried:

- Retrained model with backbone trainable to update running means etc. Similar results..Bad for batch size 1
- Retrained model using InstanceNormalization instead...Slightly better results overal but still no detections at all for batch size 1

Running out of ideas here..As I am finetuning on a relatively small dataset (~500 samples) and validating on an ever smaller one could that be the sole reason this happens...?

**Comment by mikaylagawarecki** [comment]
Hi, We use github issues only for bug reports or feature requests. Please use the forum to ask questions: https://discuss.pytorch.org/
